include/grpcpp/security/tls_credentials_options.h

```diff
@@ -0,0 +1,293 @@
+ * a credential reload request. The callback function must be called after
+ * finishing the schedule operation. See the description of the
+ * grpc_tls_credential_reload_arg struct in grpc_security.h for more details.
+ * **/
+class TlsCredentialReloadArg {
+ public:
+  /** TlsCredentialReloadArg does not take ownership of the C arg that is passed
+   * to the constructor. One must remember to free any memory allocated to the C
+   * arg after using the setter functions below. **/
+  TlsCredentialReloadArg(grpc_tls_credential_reload_arg* arg) : c_arg_(arg) {}
+  ~TlsCredentialReloadArg();
+
+  /** Getters for member fields. The callback function is not exposed.
+   * They return the corresponding fields of the underlying C arg. In the case
+   * of the key materials config, it creates a new instance of the C++ key
+   * materials config from the underlying C grpc_tls_key_materials_config. **/
+  void* cb_user_data() const;
+  std::shared_ptr<TlsKeyMaterialsConfig> key_materials_config() const;
+  grpc_ssl_certificate_config_reload_status status() const;
+  grpc::string error_details() const;
+
+  /** Setters for member fields. They modify the fields of the underlying C arg.
+   * The setters for the key_materials_config and the error_details allocate
+   * memory when modifying c_arg_, so one must remember to free c_arg_'s
+   * original key_materials_config or error_details after using the appropriate
+   * setter function.
+   * **/
+  void set_cb_user_data(void* cb_user_data);
+  void set_key_materials_config(
+      const std::shared_ptr<TlsKeyMaterialsConfig>& key_materials_config);
+  void set_status(grpc_ssl_certificate_config_reload_status status);
+  void set_error_details(const grpc::string& error_details);
+
+  /** Calls the C arg's callback function. **/
+  void OnCredentialReloadDoneCallback();
+
+ private:
+  grpc_tls_credential_reload_arg* c_arg_;
+};
+
+/** TLS credential reloag config, wraps grpc_tls_credential_reload_config. It is
+ * used for experimental purposes for now and it is subject to change.
+ *
+ * The config_user_data is read-only user data; schedule is a pointer to an
+ * application-provided callback that invokes the credential reload; cancel is a
+ * pointer to an application-provided callback that cancels a credential reload
+ * request; destruct is a pointer to an application-provided callback that
+ * cleans up any data associated to the config. See the description of the
+ * grpc_tls_credential_reload_config struct in grpc_security.h. **/
+class TlsCredentialReloadConfig {
+ public:
+  TlsCredentialReloadConfig(const void* config_user_data,
+                            int (*schedule)(void* config_user_data,
+                                            TlsCredentialReloadArg* arg),
+                            void (*cancel)(void* config_user_data,
+                                           TlsCredentialReloadArg* arg),
+                            void (*destruct)(void* config_user_data));
```

Comment:
I understand this could be related to C core, but since this is C++ API. Is it possible to use some thing like an interface that is more like C++? IMO this is not very readable.

Reply:
Do you have a good example to follow?

Reply:
I'm happy to refactor anything that you think is too C style (e.g. the function pointers?), but any examples of how this is generally done in the gRPC C++ code would be helpful. Thanks! :)

Reply:
I do not think there is a general solution for this API. However from the API surface I am curious about the proper usage (the test uses pretty simple construction and I am not sure whether that is representative). I understand this is sort of a thin wrapper of C API where the interface comes from. I am not sure about the purpose of destruct for example. If the user_data is read only as mentioned, maybe the destruct should be more like a release (unref)? 

Regarding the API, how about something like
```
struct UserConfigInterface {
  // Schedule arg on this.
  int Schedule(TlsCredentialReloadArg* arg) = 0;
  // Cancel reload with arg.
  void Cancel(TlsCredentialReloadArg* arg) = 0;
  // Config no longer needs this.
  void Release() = 0;
};
TlsCredentialReloadConfig::TlsCredentialReloadConfig(UserConfigInterface* user_data);
```
The `TlsServerAuthorizationCheckConfig` ctor may or may not share the same interface.

Reply:
Thanks for your comment and proposal. I'll refactor the API and ping you when it's ready to take a look at!

Reply:
I've refactored the credential reload and server authorization check API's, per your suggestion. I also added another test that demonstrates how exactly they will be used by the user.

Let me know what your thoughts are on the updated version, and thanks for taking a look! :)

Reply:
Thanks. This looks better. I have some more comments mostly focused on the interface.
I would suggest @yihuazhang doing another pass too.